[kimseunghyun-kr] iP

kimseunghyun-kr opened the pull request:
# Spring  

> *The overengineered Duke. Use this if you want to learn how the JAVA Spring Framework works.*  

## Notable Elements (added in this PR ✅)  

- [x] Dependency Inversion Containers  
- [x] Aspect-Oriented Programming  
- [x] JDK Proxy Behavior Injection  
- [x] Annotation-Based Programming  

## What is Dependency Inversion and Why Is It Needed?  

### Definition  
The **Dependency Inversion Principle (DIP)** states that:  
1. High-level modules should not depend on low-level modules. Both should depend on abstractions.  
2. Abstractions should not depend on details. Details should depend on abstractions.  

### Why is it needed?  
- **Reduces Coupling**: Makes modules less dependent on each other.  
- **Improves Maintainability**: Changes in low-level modules do not affect high-level modules.  
- **Enhances Testability**: Easier to mock dependencies in unit tests.  
- **Supports Dependency Injection**: Encourages the use of interfaces and abstract classes.  

### Example  

### ❌ **Without DIP (Tightly Coupled)**  
```java
class Keyboard {
    public void type() {
        System.out.println("Typing...");
    }
}

class Computer {
    private Keyboard keyboard = new Keyboard(); // Direct dependency

    public void use() {
        keyboard.type();
    }
}
```

### **With DIP (Loosely Coupled)**
```java
interface InputDevice {
    void type();
}

class Keyboard implements InputDevice {
    public void type() {
        System.out.println("Typing...");
    }
}

class Computer {
    private InputDevice inputDevice;

    public Computer(InputDevice inputDevice) { // Inject dependency
        this.inputDevice = inputDevice;
    }

    public void use() {
        inputDevice.type();
    }
}
```

### What does the DI container do?

| Responsibility                          | Description |
|-----------------------------------------|-----------------------------------------------------------|
| **Instantiation**  | Creates objects with the required dependencies injected. |
| **Object Management** | Manages object lifetimes, often as singletons or other scopes, to optimize reuse. |
| **Lifecycle Handling** | Ensures proper initialization and cleanup of objects. |

[ **more on dependency conatiners**](https://medium.com/@annuhuss/dependency-injection-container-a-simple-introduction-for-managing-objects-from-their-creation-to-cebbcb772694)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kimseunghyun-kr committed:
Added:: basic intro sequence and exit sequence, renamed file from Duke to Spring

kimseunghyun-kr committed:
refactored greeting and exit sequence into chatbotUtils. created a runtimeUtils class to reduce clutter on main

kimseunghyun-kr committed:
completed echo feature

kimseunghyun-kr committed:
created repository package for storage related things. modified BotRuntime to include add and list functionalities

kimseunghyun-kr committed:
included entity task

kimseunghyun-kr committed:
included a DIContainer to be used in future for IOC principles

kimseunghyun-kr committed:
differentiated individual commands to respective classes with a factory

kimseunghyun-kr committed:
included actionHandler to parse input, generate command and execute it

kimseunghyun-kr committed:
updated service to give required outputs

kimseunghyun-kr committed:
trailing updates on necessary changes. and minor typo fixes

kimseunghyun-kr committed:
upgraded DI container to contain a dependency graph to make initialisation independent of order of registration of classes

kimseunghyun-kr committed:
registered BotRuntime onto DI container, simplify start

kimseunghyun-kr committed:
finalized DI / AOP design choice. proxiable based on JDK proxies are used as it is really the only way. Proxiable interface implemented to enable dynamic proxies.

kimseunghyun-kr committed:
major debugging of the AOP stuff. somehow pulled it off. design compromises included

kimseunghyun-kr committed:
todo deadline event added

kimseunghyun-kr committed:
testUi

kimseunghyun-kr committed:
added delete functionality

kimseunghyun-kr committed:
update dicontainer to allow spring-like multi-interface inheritance

kimseunghyun-kr committed:
update AOP to reflect changes in proxyenable and multi-proxyenable inheritance

kimseunghyun-kr committed:
update entity with UUID to allow better tracking in repository

kimseunghyun-kr committed:
minor typo fix in Actions enums

kimseunghyun-kr committed:
add entity managing base code, incorporate scheduler within taskflusher for more efficient saves

kimseunghyun-kr committed:
add CRUDrepository for base CRUD repository outline, add filebackedTaskRepository to facilitate File I/O base buffer , write, read, modified ItaskRepository to be able to show size

kimseunghyun-kr committed:
add flush scheduler start and stop within the botruntime, minor bugfix of 'add' causing error

kimseunghyun-kr committed:
add markdirty and additional dependency to new interface for file buffer -> note that this has now become strongly coupled to the individual repositories -> may be best to have a separate service layer managing this in the future

kimseunghyun-kr committed:
util updated to include serialiser and deserialiser aspect for the file saves. directoryinitialize utils to initialize directory

kimseunghyun-kr committed:
update main to include all the changed aspects to dependency container

kimseunghyun-kr committed:
add command for date search and servicecoordinator to handle double repository issue

kimseunghyun-kr committed:
add command for date search and servicecoordinator to handle double repository issue

kimseunghyun-kr committed:
add command for date search and servicecoordinator to handle double repository issue

kimseunghyun-kr committed:
Merge branch 'gradle'
merge gradle into master for gradle support

kimseunghyun-kr committed:
created mocks and tests for mockTaskService, commandFactory

kimseunghyun-kr committed:
jar

kimseunghyun-kr committed:
Create artifact jar : for pre alpha releasereate artifact jar : for pre

kimseunghyun-kr committed:
Create artifact jar : pre alpha release

kimseunghyun-kr committed:
Refactor code : include javadocs for more methods

kimseunghyun-kr committed:
Add javadocs for half of methods

kimseunghyun-kr committed:
add javadocs for half of the methods

kimseunghyun-kr committed:
coding standard edit

kimseunghyun-kr committed:
Edit to meet coding standard

kimseunghyun-kr committed:
add replay-log feature for partial edit during saves to save I/O, increase autosave period to 1 minute

kimseunghyun-kr committed:
Add replay-log for partial edit during save

JohannsenLum commented on src/main/java/util/CommandMapper.java:
```diff
@@ -0,0 +1,119 @@
+        COMMAND_MAPPINGS.put("bye", Actions.TERMINATE);
+    }
+
+    /**
+     * Maps a user-provided command string to an {@link Actions} enum.
+     *
+     * @param command The user input command.
+     * @return The corresponding {@link Actions} enum.
+     * @throws UserFacingException If the command is unknown or deprecated.
+     */
```
Well documented javaDocs here, very informative!

JohannsenLum reviewed (commented):
Great job on your IP! Your approach is clean, well structured and follows good practices. I especially appreciate how you structured the different classes and did more than you were tasked to. Keep up the great work!

kimseunghyun-kr committed:
Add feat: included a new bash script to doctor commit history

kimseunghyun-kr committed:
Add feat: include GUI support
Edit feat: refactor code for static setter injection

JohannsenLum commented on src/main/java/Spring.java:
```diff
@@ -0,0 +1,46 @@
+import runtime.IBotRunTime;
+import util.DirectoryInitializeUtils;
+
+
```
Class header

JohannsenLum commented on src/main/java/Spring.java:
```diff
@@ -0,0 +1,46 @@
+
+public class Spring {
+    public static void main(String[] args) {
+        DependencyInjectionContainer container = new DependencyInjectionContainer();
```
private modifier preferred

JohannsenLum commented on src/main/java/controller/ControllerResponse.java:
```diff
@@ -0,0 +1,30 @@
+package controller;
+
```
class header

JohannsenLum commented on src/main/java/dispatcher/CliDispatcher.java:
```diff
@@ -0,0 +1,46 @@
+import runtime.ActionHandler;
+import runtime.CommandDao;
+
+public class CliDispatcher implements IDispatcher {
```
class header

JohannsenLum reviewed (commented):
Did a second round of review and found it pretty good! Mainly a few headers missing that would be good to have.

kimseunghyun-kr committed:
Add feat: minor syntax typo updates

kimseunghyun-kr committed:
Add feat: clear command to clear all tasks

kimseunghyun-kr committed:
Add feat: clear command to clear all tasks

kimseunghyun-kr committed:
Merge branch 'branch-Level-10'

kimseunghyun-kr committed:
Edit feat: solved merge overlooked merge conflict

kimseunghyun-kr committed:
Edit feat: solved merge overlooked merge conflict

kimseunghyun-kr committed:
Edit : change autosave period from 60 seconds to 30

kimseunghyun-kr committed:
Edit : change autosave period from 60 seconds to 30

noahang commented on src/main/java/entity/tasks/DeadLine.java:
```diff
@@ -0,0 +1,34 @@
+
+public class DeadLine extends Task {
+
+    private LocalDateTime dueby;
```
should use camelCase for `dueby` ?

noahang commented on src/main/java/entity/tasks/Events.java:
```diff
@@ -0,0 +1,44 @@
+
+public class Events extends Task {
+
+    private LocalDateTime startat;
```
I think camelCase should be used for these members as well,  I noticed this issue in several other places too!

noahang commented on src/main/java/entity/tasks/Events.java:
```diff
@@ -0,0 +1,44 @@
+
+import java.time.LocalDateTime;
+
+
```
missing header comment, I noticed this issue in most of your other classes as well.

noahang commented on src/main/java/gui/components/DialogBox.java:
```diff
@@ -0,0 +1,72 @@
+        dialog.setText(text);
+        dialog.setMinHeight(Math.max(60, (dialog.getFont().getSize() + 10) * countLines(text)));
+    }
+
```
missing header comment, I noticed this issue in other methods as well

noahang reviewed (commented):
LGTM other than some minor style issues. I'm guessing you aren't very concerned with javadocs for now given that your classes will likely undergo some changes down the line, and that's fine. Overall, well done! It is very cool to see a different framework be used for the ip.

kimseunghyun-kr committed:
Edit : re-included annotationprocessor settings for lombok
which got lost during branch rebase

kimseunghyun-kr committed:
Edit : refactored code style based on linter

kimseunghyun-kr committed:
Edit : changed else return value within getUpdatedTaskType
in TaskDialogCotainer from null to current value
This will allow greater idempotency

Edit : moved up branchDoctor.sh from /src to project root

kimseunghyun-kr committed:
Add: Update and Interactive update feature.

Interface to unify the CLI and the GUI variants
under interactiveexecutionservice package

GUI has been tested and working
CLI not in the clear yet.

Move: DAO into dao within service package

Move: command executing logic from dispatcher/Mainwindow
to CommandExecutionService to handle interactive updates
better under the MVC principle.

Add: Lombok integration within the tasks and subclasses
replaced getter setters with lombok

Move: static variable setter injection from clistartup()
to FxmlStaticSetterInjectionConfig

kimseunghyun-kr committed:
Add: @NonNull lombok in place fior assertions
on locations that needs to have null checks.
This is so as the @NonNull essentially does an assert
under the hood

kimseunghyun-kr committed:
Edit: included javadocs for all main classes.
this is to adhere to the coding quality standareds

Edit: modified Dicontainer's boolean initialized to isinitialized
this is to adhere to code naming standards

kimseunghyun-kr committed:
Edit : refactored code style based on linter

kimseunghyun-kr committed:
Add: Update and Interactive update feature

Interface to unify the CLI and the GUI variants
under interactiveexecutionservice package

GUI has been tested and working
CLI not in the clear yet.

Move: DAO into dao within service package

Move: command executing logic from dispatcher/Mainwindow
to CommandExecutionService to handle interactive updates
better under the MVC principle.

Add: Lombok integration within the tasks and subclasses
replaced getter setters with lombok

Move: static variable setter injection from clistartup()
to FxmlStaticSetterInjectionConfig

kimseunghyun-kr committed:
Fix : error for the update code portion

kimseunghyun-kr committed:
Edit feat: solved overlooked merge conflict

kimseunghyun-kr committed:
Fix : included branchdoctor.sh into gitignore

kimseunghyun-kr committed:
Edit : update TaskDialogController if else conditions
and return values to be idempotent
This would lalow idempotent inputs, no changes, to
work seamlessly

kimseunghyun-kr committed:
Merge remote-tracking branch 'origin/branch-A-CodeQuality'
This is done due to master rebase issue

kimseunghyun-kr committed:
Merge remote-tracking branch 'origin/branch-A-Assertions'
This has been done to signal merge to ip autochecker
due to history changes on master causing diverge

kimseunghyun-kr committed:
Edit: refactored code and changed names to adhere to
coding standards

kimseunghyun-kr committed:
Add: better catches for exception handling

kimseunghyun-kr committed:
Add : readme.md user docs

Edit: minor bug fixes

kimseunghyun-kr committed:
Add : UI for product showcase

kimseunghyun-kr committed:
Edit : pages to reformat table and include picture

kimseunghyun-kr committed:
Add : tests for better testing on TaskService

kimseunghyun-kr committed:
Add : _config.yml file to ensure the github pages use
GFM to render and not kramdown to prevent tables from
breaking

kimseunghyun-kr committed:
Edit : newline between the heading and table for
github pages render

kimseunghyun-kr committed:
Edit : UI.png to Ui.png in order to be scanned

kimseunghyun-kr committed:
Add : explanation stating task can only be one word

kimseunghyun-kr committed:
Add : restore setting title back

kimseunghyun-kr committed:
Add : modify Ui.png to match refspec

kimseunghyun-kr committed:
Add : make recovery more robust

kimseunghyun-kr committed:
Add : make recovery more robust, documented

kimseunghyun-kr committed:
Add : make recovery more robust, backup
now merges the content with initial scan
